std/machines/arith.asm

```diff
@@ -155,24 +155,11 @@ machine Arith with
         _ => Query::None
     };
 
-    col witness y1_0(i) query hint_if_eq0(quotient_hint, 0);
-    col witness y1_1(i) query hint_if_eq0(quotient_hint, 1);
-    col witness y1_2(i) query hint_if_eq0(quotient_hint, 2);
-    col witness y1_3(i) query hint_if_eq0(quotient_hint, 3);
-    col witness y1_4(i) query hint_if_eq0(quotient_hint, 4);
-    col witness y1_5(i) query hint_if_eq0(quotient_hint, 5);
-    col witness y1_6(i) query hint_if_eq0(quotient_hint, 6);
-    col witness y1_7(i) query hint_if_eq0(quotient_hint, 7);
-    col witness y1_8(i) query hint_if_eq0(quotient_hint, 8);
-    col witness y1_9(i) query hint_if_eq0(quotient_hint, 9);
-    col witness y1_10(i) query hint_if_eq0(quotient_hint, 10);
-    col witness y1_11(i) query hint_if_eq0(quotient_hint, 11);
-    col witness y1_12(i) query hint_if_eq0(quotient_hint, 12);
-    col witness y1_13(i) query hint_if_eq0(quotient_hint, 13);
-    col witness y1_14(i) query hint_if_eq0(quotient_hint, 14);
-    col witness y1_15(i) query hint_if_eq0(quotient_hint, 15);
-
-    let y1: expr[] = [y1_0, y1_1, y1_2, y1_3, y1_4, y1_5, y1_6, y1_7, y1_8, y1_9, y1_10, y1_11, y1_12, y1_13, y1_14, y1_15];
+    let y1: inter[] = array::new(16, |i| {
```

Comment:
If this is changed to `expr[]` (as it should), the query processor does not work properly any more.

The reason is that in constraints, all references to `y1` are replaced by the cached evaluated value for y1. In other words, a single reference to y1 causes y1 to be "computed", but the array::new function is only executed once.

Query functions, though, are not really processed by the condenser. This leads to the problem that when evaluating the query, y1 is evaluated again and it tries to create a witness column which of course fails. I'm not sure how to properly fix this. Maybe the `prover` section will solve this.

Reply:
Created an issue: https://github.com/powdr-labs/powdr/issues/1713